specification/metrics/semantic_conventions/system-metrics.md

```diff
@@ -33,12 +33,14 @@ instruments not explicitly defined in the specification.
-| system.cpu.time        |                                                                                                          | s     | Counter                                           | Double     | state            | idle, user, system, interrupt, etc. |
-|                        |                                                                                                          |       |                                                   |            | cpu              | CPU number [0..n-1]                 |
-| system.cpu.utilization | Difference in system.cpu.time since the last measurement, divided by the elapsed time and number of CPUs | 1     | Gauge                                             | Double     | state            | idle, user, system, interrupt, etc. |
-|                        |                                                                                                          |       |                                                   |            | cpu              | CPU number (0..n)                   |
+| Name                      | Description                                                                                              | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s) | Attribute Values                    |
+| ------------------------- | -------------------------------------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ---------------- | ----------------------------------- |
+| system.cpu.time           |                                                                                                          | s     | Counter                                           | Double     | state            | idle, user, system, interrupt, etc. |
+|                           |                                                                                                          |       |                                                   |            | cpu              | CPU number [0..n-1]                 |
+| system.cpu.utilization    | Difference in system.cpu.time since the last measurement, divided by the elapsed time and number of CPUs | 1     | Gauge                                             | Double     | state            | idle, user, system, interrupt, etc. |
+|                           |                                                                                                          |       |                                                   |            | cpu              | CPU number (0..n)                   |
+| system.cpu.count.physical | Reports the number of available physical CPUs - may changes over time                                | 1     | UpDownCounters                                    | Int64      | device           |                                     |
```

Comment:
Would that make sense to add the `physical`/`virtual` suffixes as attribute values instead?
Also what are the possible attribute values for the `device` Attribute Key?

Reply:
What's the use case for splitting up physical/virtual? Additionally, is it possible for instrumentation to reliably detect if the CPU is physical or virtual? My initial thoughts are I think I would prefer it if this was just one metric.

Reply:
> What's the use case for splitting up physical/virtual? 

At a minimum, we should clarify which one of the two we are adding as a semantic convention, since the term can be ambiguous. I think both are helpful to understand the hardware a machine has, in our (Datadog's) case, it would be helpful to have both to provide feature parity with our existing vendor-specific instrumentation.

>  is it possible for instrumentation to reliably detect if the CPU is physical or virtual?

Widely used libraries such as psutil allow [fetching this information](https://psutil.readthedocs.io/en/latest/#psutil.cpu_count) on all major operating systems (Linux, Windows, macOS, *BSD, AIX...). For example, on Linux it is possible to fetch this information so long as the process has access to `/proc/cpuinfo` and `/sys/devices/`. This is a similar level of access to that required for fetching the information related to other system metrics already in the spec.

Reply:
I think we need a definition of (or a link to) what "virtual CPU count" is. The link to psutil talks about "logical" cpu count, not virtual.

Are we trying to capture the nuances of virtualized environments? For a system running on a VM what does each of these metrics return? Is "physical" count the actual physical count or the number allocated by the hypervisor to the VM?

Reply:
We started talking about "physical" vs "logical" on the original discussion (https://github.com/open-telemetry/semantic-conventions/pull/99#discussion_r1228223112 ) but ended talking about physical vs virtual and I failed to realize that this was confusing and could be interpreted as virtual in the sense of a VM.

To me, **the important distinction is between physical** (number of CPUs available in the machine) **and logical** (physical multiplied by the number of threads that can run on each CPU through hyperthreading). This is the way `psutil` defines it:
>  “logical CPUs” means the number of physical cores multiplied by the number of threads that can run on each core (this is known as Hyper Threading)

see also [this SE question](https://unix.stackexchange.com/a/88290) or the Wikipedia article on [Hyperthreading](https://en.wikipedia.org/wiki/Hyper-threading).

The terms "virtual" and "logical" are used interchangeably sometimes in this context (e.g. see the first section on the Wikipedia article linked above) but virtual can also mean in a VM in common parlance, so based on the above discussion I feel like logical is a less ambiguous term.

On a virtualized environment the 'physical' number may not correspond to the actual number of CPUs the machine has depending on what the hypervisor allocated, but I still think physical is the common name used for this. 

Since the physical and logical definitions do not depend on the hypervisor I think it may be clearer now that these numbers are available to instrumentation provided they can read the appropriate system files/call the right APIs.

@tigrannajaryan @jamesmoessis does this make the definitions clearer to you?

Reply:
I think we should avoid using "virtual" and stick to "logical". The word "virtual" appears only twice on wikipedia article and "logical" is mentioned 15 times. That will help avoid confusion for virtualized environments.

If we stick to "physical" and "logical" I don't think we need to strictly define the behavior for VMs. The numbers should be "whatever the guest OS sees".

Reply:
I believe the most common use case for these metrics on the backends would be to calculate cpu utilization from `system.cpu.time` metric. We should make it clear which one of them should be used for that purpose.

Reply:
To add to what @dmitryax said. I think we need to be clear which of these new metrics' count is supposed to be the total number of values you see for `cpu` attribute of `system.cpu.time` metric. There must be an exact correspondence conceptually between what `system.cpu.time`/`cpu` dimension describes and what one of these 2 new metrics describes.

Reply:
`system.cpu.time`, at least on psutil/gopsutil and (therefore in the current implementation of the hostmetrics receiver) reports the time for each **logical** CPU:

<details>
<summary>
Go and Python short programs showing that the logical CPUs are used for <code>system.cpu.time</code>
</summary>

Short check using `psutil`:

```python
#! /usr/bin/python
import psutil
times = psutil.cpu_times(percpu=True)
logical_count = psutil.cpu_count(logical=True)

print(f"logical count: {logical_count} system.cpu.time number of CPUs: {len(times)}")
```

**Output:**
```
❯ ./main.py
logical count: 20 system.cpu.time number of CPUs: 20
```

Go program using `gopsutil` (the library used by the host metrics receiver):

```go
package main

import (
	"fmt"
	"github.com/shirou/gopsutil/v3/cpu"
)

func main(){
       // fn used by hostmetrics receiver here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3dfc0f9/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go#L39
	times, err := cpu.Times(true)
	if err != nil {
		panic(err)
	}

	logicalCount, err := cpu.Counts(true)
	if err != nil {
		panic(err)
	}
	
	fmt.Printf("logical count: %d system.cpu.time number of CPUs: %d\n", logicalCount, len(times))
}
```

**Output:**
```
❯ go run main.go
logical count: 20 system.cpu.time number of CPUs: 20
```
</details>

We can clarify this on the description of `system.cpu.time`.